File: app/backend/schemas/websocket_messages.py
from pydantic import BaseModel, Field
from typing import Any, Literal, Dict

class WebSocketMessageBase(BaseModel):
    job_id: int
    type: str # Differentiates message types, e.g., 'status', 'partial_result', 'final_result', 'error'
    data: Dict[str, Any] | None = None # Flexible data payload

class WebSocketStatusUpdate(WebSocketMessageBase):
    type: Literal["status_update"] = "status_update"
    status: str # e.g., "PENDING", "PROCESSING_EXPLANATION", "GENERATING_CODE", "COMPLETED", "FAILED"
    message: str | None = None # Optional descriptive message for the status

class WebSocketPartialResult(WebSocketMessageBase):
    type: Literal["partial_result"] = "partial_result"
    result_type: Literal[
        "explanation", 
        "concepts", 
        "storyboard", 
        "enhanced_storyboard", 
        "generated_code", 
        "review_comments"
    ]
    content: str # The actual partial result content

class WebSocketFinalResult(WebSocketMessageBase):
    type: Literal["final_result"] = "final_result"
    refactored_code: str | None = None
    video_url: str | None = None # For Manim video link
    manim_error: str | None = None # For any errors during Manim processing
    message: str = "Processing completed successfully."

class WebSocketError(WebSocketMessageBase):
    type: Literal["error"] = "error"
    error_message: str

# For messages from client to server over WebSocket (if any, besides initial connect)
class WebSocketClientMessage(BaseModel):
    action: str # e.g., 'start_processing', 'subscribe_updates'
    payload: Dict[str, Any] | None = None 

File: app/frontend/src/types/websocketMessages.ts
export interface WebSocketMessageBase {
    job_id: number;
    type: string;
    data?: Record<string, any>;
}

export interface WebSocketStatusUpdateMessage extends WebSocketMessageBase {
    type: "status_update";
    status: string; // e.g., "PENDING", "PROCESSING_EXPLANATION", "COMPLETED", "FAILED"
    message?: string;
}

export type PartialResultType = 
    | "explanation"
    | "concepts"
    | "storyboard"
    | "enhanced_storyboard"
    | "generated_code"
    | "review_comments";

export interface WebSocketPartialResultMessage extends WebSocketMessageBase {
    type: "partial_result";
    result_type: PartialResultType;
    content: string;
}

export interface WebSocketFinalResultMessage extends WebSocketMessageBase {
    type: "final_result";
    refactored_code?: string | null;
    video_url?: string | null;
    manim_error?: string | null;
    message?: string;
}

export interface WebSocketErrorMessage extends WebSocketMessageBase {
    type: "error";
    error_message: string;
}

export type AgentWebSocketMessage = 
    | WebSocketStatusUpdateMessage 
    | WebSocketPartialResultMessage 
    | WebSocketFinalResultMessage 
    | WebSocketErrorMessage; 